Source/Artifacts/Artifacts.cs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reactive.Subjects;
+using Microsoft.Extensions.Logging;
+
+namespace Dolittle.SDK.Artifacts
+{
+    /// <summary>
+    /// Represents an implementation of <see cref="IArtifacts" />.
+    /// </summary>
+    public class Artifacts : IArtifacts, IDisposable
+    {
+        readonly ReplaySubject<ArtifactAssociation> _registered;
+        readonly BehaviorSubject<IDictionary<Type, Artifact>> _associations;
+        readonly ILogger<Artifacts> _logger;
+        bool _disposed;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Artifacts"/> class.
+        /// </summary>
+        /// <param name="associations">The <see cref="IDictionary{TKey, TValue}"> artifact associations </see>.</param>
+        /// <param name="logger">The <see cref="ILogger" />.</param>
+        public Artifacts(IDictionary<Type, Artifact> associations, ILogger<Artifacts> logger)
+        {
+            _registered = new ReplaySubject<ArtifactAssociation>();
+            _associations = new BehaviorSubject<IDictionary<Type, Artifact>>(new Dictionary<Type, Artifact>());
+            _logger = logger;
+
+            foreach ((var type, var artifact) in associations)
+            {
+                _registered.OnNext(new ArtifactAssociation(type, artifact));
+            }
+
+            _registered.Subscribe(AddAssociation);
+        }
```

Comment:
Maybe cleaner to call Associate from the outside (the builder) and just have the subscription set up here?

Reply:
Also - no need to explicitly use the generic logger - that's just for creating new instances. This thing doesn't really care about that.

Reply:
Isn't it best for a class to own a ILogger for its own Type? To make sure that the logger it uses doesn't print anything weird, for instance using another class (category)

Reply:
> Maybe cleaner to call Associate from the outside (the builder) and just have the subscription set up here?

Sure we can do that. Again just mirrored the JS SDK implementation